rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+  const publicUrl = getPublicAddress();
+  const absoluteUrl = joinRemovingDuplicateAndTrailingSlash(
+    publicUrl,
+    // note: this is actually wrong. We would need the `requestScopePath` here
+    // as this currently returns `${this.serverBasePath}${requestScopePath}` and the basePath
+    // is already included in `publicAddress`
+    serverContract.basePath.get(request), 
+    path
+  );
+}
+```
+
+Search results will then be consolidated before being returned to convert relative urls to absolute ones:
+
+```ts
+const consolidateResult(result: GlobalSearchResult, request: KibanaRequest) {
+  if(isUrlPath(result.url)) {
+    result.url = http.getAbsoluteUrl(result.url, request)
+  }
+}
+```
+
+#### Redirecting to a result
+
+Having absolute urls in GS results is a necessity for external consumers, and makes the API more consistent than mixing
+relative and absolute urls, however this makes it less trivial for UI consumers to redirect to a given result in a SPA
+friendly way (using `application.navigateTo` instead of triggering a full page refresh).
+
+This is why `NavigableGlobalSearchResult.navigateTo` has been introduced, to let `core` handles this navigation logic.
+
+When using `navigateTo` from a result instance, the following logic will be executed:
+
+If all 3 of these criteria are true for `result.url`:
+- The domain of the URL  matches the domain of the `publicUrl`
+- The pathname of the URL starts with the current basePath (eg. /mybasepath/s/my-space)
+- The pathname segment after the basePath matches an application route (eg. /app/<id>/)
+
+Then: match the pathname segment to the corresponding application and do the SPA navigation to that application using 
+the remaining pathname segment
+
+Otherwise: do a full page navigation (`window.location.assign()`)
+
+### searching from the server side
+
+When calling `GlobalSearchServiceStart.find` from the server-side service:
+
+- the service will call `find` on each server-side registered result provider and collect the resulting result observables
+
+- then, the service will merge every result observable and trigger the next step on every emission until either
+    - A predefined timeout duration is reached
+    - All result observables are completed
+   
+- on every emission of the merged observable, the results will be aggregated to the existing list and sorted following 
+the logic defined in the [results aggregation](#results-aggregation) section
+
+A very naive implementation of this behavior would be:
+
+```ts
+search(
+  term: string,
+  options: GlobalSearchOptions,
+  request: KibanaRequest
+): Observable<GlobalSearchResponse> {
+  const fromProviders$ = this.providers.map(p =>
+    p.find(term, options, contextFromRequest(request))
+  );
+  const results: GlobalSearchResult[] = [];
+  return merge([...fromProviders$]).pipe(
+    takeUntil(timeout$),
+    map(newResults => {
+      results.push(...newResults);
+      return order(results);
+    }),
+  );
```

Comment:
TBH I'm still mixed on whether we should return all or only new results here. 

If we want to handle the sorting of the results, we have no other choice than to return them all every time. Also it's probably way more easy to consume an up-to-date list than re-aggregating on their side on every emission.

Maybe we could add an option in `GlobalSearchOptions` at a later time to only emit the (unsorted) new results instead? Not sure of the need.

Reply:
Updating results does have the problem of results reordering out from under you... Not sure how we want to handle that on the UI? @ryankeairns

Not doing it this way means slower results but doing it this way means sometimes a frustrating UX. We could add all new results to the bottom or something but that might lead to weird list orders...

Reply:
I'm having a hard time envisioning the alternatives... I'm tempted to think I would accept the slower process vs having the results change unexpectedly, but it depends on how slow :) Is the 'show all results' approach one we can start with without hindering future options? I sense there is an iterative approach to take here that allows us to address scaling issues as this grows.

Reply:
I think returning only new results here give us a lot more flexibility in how the results are consumed. It would allow us to do something like wait 500ms for as many results as possible, and then add all new results to the end. It would also allow us to support global search sort customizations, without affecting all other usages.

Reply:
@kobelb @joshdover I don't have any issue with returning only new results. However I just want to be sure that everyone understand the consequences.

 By choosing this option:
- We let the consumers aggregates the results on every emission on their side (even if trivial with some rxjs operators)
- More important: we no longer can handle sorting the results, and we let this responsibility to the consumers that would need to implements this logic based on their own arbitrary rules and on the `score` value of each result.

Is this alright?

Reply:
> Is this alright?

It sounds reasonable to me. If we end up needing this code in multiple places, we can generalize it at that time.

Reply:
Unless additional arguments are presented, I'm leaning for the batched results here instead of aggregated results.

Reply:
done in https://github.com/elastic/kibana/pull/64284/commits/d2289f79e44ac6d042a049ee307395b25df5913b